### PR TITLE
refactor to use es6 classes, async/await and request-promise-native

### PR DESCRIPTION
**Pull request by jwickens**
* use class instead of an emitter singleton
* use await / async
* use request-promise-native

### CODE REVIEW

**Comment by jwickens** [comment]
Hey @bubenshchykov before i submit this as a PR just had a few questions about how to proceed:

A) "ui url"
The uiUrl parsed here - https://github.com/bubenshchykov/ngrok/blob/4a6d1c656ee1bd825a17e6801b3398ebdaaf2136/index.js#L188

What is that? How should we return it as on here in this gist i return just the publicUrl when you await ngrok.connect the alternative i thought of was to return an object instead but that doesnt seem really neat.

B) auth token
The authToken method here https://gist.github.com/jwickens/0a6dc0655f756d01d01f7057537b057b#file-index-js-L146

I couldnt figure out in the original code how this works, how the auth token makes it back in. In fact this method is never called by the connect method here in the gist.

(C) node.js version supported

Please note that request-promise-native requires node greater than 0.12 (https://github.com/request/request-promise-native/blob/1874877850a59152915c9e9cbacbdc577486cca5/package.json#L33)

The async/await, classes and destructuring runs fine natively on node.js >= 8.3

For versions below that babel can be used, however that comes at the cost of making the code less easily inspectable from node_modules depending on how low you want to go.

So what do you want to target?

**Comment by toverux** [comment]
@jwickens If your PR gets merged, would you be able to update the [TypeScript definitions](https://github.com/bubenshchykov/ngrok/blob/master/ngrok.d.ts) (if necessary, I didn't take a close look to the changes)? If not, I can handle it.

**Comment by bubenshchykov** [comment]
C) I think we should just target node > 8.3, as users of older node can use older versions of this module
B) authtoken always sits in ngrok config, it's stored there and retrieved automatically by ngrok when someone creates tunnel. So our task is just to push it to config via ```ngrok authtoken ``` command
A) uiUrl - I guess it's ngrok internal api (and UI) url. Have you seen ngrok UI? It's where you can manage and see history of requests. uiUrl was exposed as third return parameter of connect function here https://github.com/bubenshchykov/ngrok/pull/73, but I never personally used it myself :)

**Comment by bubenshchykov** [comment]
offtopic: the test suite for ngrok is great, it uses free and paid authtokens to ensure everything works. I had them open public, but people has started to use this authtokens for their projects, so tests turned red because of usage restrictions. I made them hidden in travis, but it makes it impossible for contributors to run tests :( Not sure how to handle it nicer

**Comment by bubenshchykov** [comment]
@toverux @jwickens guys, what's the best approach of not breaking the interfaces and switching to promises/async/await (instead of callbacks in interface functions)? I guess by accepting this PR everyone will be forced to use promises/async/await (with latest node), as supporting it together with callback will be hell. Do you think it's fine to set a major version update and say "use that" ? :)

**Comment by jwickens** [comment]
Really its up to you and the other ngrok maintainers @bubenshchykov - my personal opinion is that yes its fine. Sorry BTW ive been slow on the cleaning the other bits here -- will get to it soon!

**Comment by toverux** [comment]
I think it's fine. This library is not suffering of having too much breaking changes and it's 2018, a major version for promises support looks reasonable.
I tend to wrap anything that's callback-style to get promises today. So I'm very happy when a library moves to promises. And we don't have the complexity, inertia and documentation amount of projects like (for example) mongoose, that is supporting both promises and callbacks.

**Comment by jwickens** [comment]
Finally had some time to finish up changes to package.json and update some tests for the guest access. As a cheeky aside, maybe if someone from ngrok wanted to reward us with a plan I could do the tests involving authtoken 😉. In any case maintainers should have access to push changes to that branch to help finish those tests.

@toverux yes I think the typescript definitions will need updating, as all the methods now return a promise rather than using callbacks. In addition the main export is a class rather than a singleton. I would do it but am not 100% comfortable with typescript.

@bubenshchykov Regarding the UI URL, wow Ive used the nrgok binary standalone so many times and I never knew I could open up localhost:4040 and get that UI. Cool! Is port 4040 always the port chosen?

In the tests I updated, besides ongoing uiUrl stuff, I had to skip the error event being thrown on connect. I'm not really sure if it makes much sense to emit an error event on calling any of the methods, as that would involve a lot of try/catch blocks and really anyone calling those methods should be catching the errors when they call them.

To summarise, three items remaining probably from my point of view

* sort UI URL
* deprecate (or not) the error event
* update remaining tests

**Comment by bubenshchykov** [comment]
@jwickens awesome!

1) tests - still don't know how to set tokens, but protect their usage to travis env in PRs only. again, all tokens were public, but people used them for work projects, so tunnels in tests were dying because of quota limits hit )))

2)  I think to kill event emitter at all. I guess it was added in one PR because it was the only option to write "sync" code :) With this PR, people can do with promises anything they want, with or without await, right?

3) uiURL. even though it's useful, I don't want promises to return object instead of simple "tunnel url". Since uiURL is always the same per ngrok process, we can move it to separate static method ngrok.getCurrentApiUrl() // or getCurrentUIUrl

**Comment by bubenshchykov** [comment]
Hey @jwickens. FYI, I took your wonderful PR, made all tests green and split the code a bit here
https://github.com/bubenshchykov/ngrok/pull/108

I still need to clean it up and add eslint, but so it's stable so far.
I'm closing this PR and will ping you when async-await gets merged!

**Comment by bubenshchykov** [comment]
Hi guys, I pushed async-await to master
@toverux could you please create a PR for type definitions
I tried to use some tools to generate d.ts, but had no luck :)
After that, I'll bump this huge breaking change as 3.0.0 to npm yaaaai!
